### PR TITLE
version-hash command for mint-client-cli and mint-rpc-client

### PR DESCRIPTION
**Pull request by NicolaLS**
I decided to also do some refactoring to make the CLI logic cleaner and also give the option to work without workdir.

This is a step towards #555 the other binaries that need to be changed are:
1. clientd
2. clientd-cli
3. configgen
4. fedimintd
5. ln-gateway

I think it is best if I do each of them in a individual PR.

### CODE REVIEW

**Comment by dpc** [comment]
Don't fail if the file is not available. When it's not available hardcode some fake and unique string instead. I'm going to add additional "branding" step in Nix that will replace that hardcoded string with actual hash somehow as the very last step.

File: client/cli/build.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+use std::{env, fs::File, process::Command};
+fn main() {
+    let prev_hash_var = env::var("GIT_HASH");
+    if let Ok(prev_hash) = prev_hash_var {
+        if prev_hash.len() < 4 || &prev_hash[0..4] != "fedi" {
```
Comment (dpc) [inline client/cli/build.rs:5]: I don't understand this one. Users that want to override the hash should be able to just set any, no?

Possibly we should name the variable itself `FEDIMINT_GIT_HASH` if we are afraid of collisions with other software, etc.

File: client/cli/src/bin/mint-rpc-client.rs [hunk 2]
```diff
@@ -1,22 +1,48 @@
...
-    let call = ApiCall::parse();
-    let arg: serde_json::Value = serde_json::from_str(&call.arg).unwrap();
-    let api = WsFederationApi::new(0, vec![(PeerId::from(0), call.url)]);
-    let response: serde_json::Value = api.request(&call.method, arg).await.unwrap();
-    let formatted = serde_json::to_string_pretty(&response).unwrap();
-    print!("{}", formatted);
+    let cli = Cli::parse();
+    match cli.command {
+        Commands::VersionHash => {
+            if let Ok(hash) = env::var("GIT_HASH") {
```
Comment (dpc) [inline client/cli/src/bin/mint-rpc-client.rs:34]: In the app itself, we should only use the compiled time one only, no? `env!("GIT_HASH")`

File: client/cli/src/bin/mint-rpc-client.rs [hunk 3]
```diff
@@ -1,22 +1,48 @@
...
+                println!("{}", hash);
```
Comment (dpc) [inline client/cli/src/bin/mint-rpc-client.rs:35]: Style: I would just inline `hash`.

File: client/cli/build.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        }
+    }
+    let git_hash = if File::open("./../../.git/HEAD").is_ok() {
+        let output = Command::new("git")
+            .args(&["rev-parse", "HEAD"])
+            .output()
+            .unwrap();
+        String::from_utf8(output.stdout).unwrap()
+    } else {
+        String::from("fedif7069afbe457afa1d2683a099c7af48a523c")
```
Comment (dpc) [inline client/cli/build.rs:18]: Some comment about why is this hardcoded for the confused passer-by would be good to have.

Also we only want something catchy in here (visually), it doesn't have to be `fedi` related.

File: client/cli/src/bin/mint-rpc-client.rs (hunk 3 above)
Comment (NicolaLS) [reply client/cli/src/bin/mint-rpc-client.rs:35]: you mean `println!("{}, env!("GIT_HASH"))` ?

File: scripts/build.sh [hunk 5]
```diff
@@ -33,7 +33,7 @@ $FM_BIN_DIR/configgen --out-dir $FM_CFG_DIR --num-nodes $FM_FED_SIZE --hbbft-bas
 export FM_LN1="lightning-cli --network regtest --lightning-dir=$FM_LN1_DIR"
 export FM_LN2="lightning-cli --network regtest --lightning-dir=$FM_LN2_DIR"
 export FM_BTC_CLIENT="bitcoin-cli -regtest -rpcuser=bitcoin -rpcpassword=bitcoin"
-export FM_MINT_CLIENT="$FM_BIN_DIR/mint-client-cli $FM_CFG_DIR"
```
Comment (dpc) [inline scripts/build.sh:36]: Is this a fix? Probably should make it another PR.

**Review by dpc — COMMENTED** [review]
One comment, rest LGTM

File: scripts/build.sh (hunk 5 above)
Comment (NicolaLS) [reply scripts/build.sh:36]: I changed the `mint-client-cli` to not take the workdir as an argument but as a flag so other commands which dont need the workdir are not bothered by it

File: client/cli/src/main.rs [hunk 6]
```diff
@@ -280,15 +289,38 @@ struct PayRequest {
...
+            "{}",
+            CliOutput::VersionHash {
+                hash: env!("GIT_HASH").to_string()
+            }
+        );
+        return;
+    };
+    //If we don't make the workdir optional future commands that might not need it will be
+    //bothered by that
+    let workdir = if let Some(workdir) = &cli.workdir {
```
Comment (dpc) [inline client/cli/src/main.rs:312]: This part is not great.

I see couple of options/combinations of:

* Prepare a separate `#[derive(Parser)] struct CliNoWorkdir` that would be a subset of `Cli` containing only commands that don't need the workdir, and basically use `try_parse` to fall back between them.
* Parse `std::env::args_os` manually to handle this special case.
* Move workdir to be an argumento every subcommand that needs it. (kind of the cleanest way, but also most involved).
* See if https://docs.rs/clap/latest/clap/builder/struct.Arg.html#method.requires can be addded on all commands that require `workdir`.

Comment (NicolaLS) [reply client/cli/src/main.rs:312]: I think doing the other `Parser` might look clumsier but is cleaner than handling it with `std::env::args_os` directly. Good Idea

Comment (NicolaLS) [reply client/cli/src/main.rs:312]: yes "Any [Arg method](https://docs.rs/clap/latest/clap/builder/struct.Arg.html) can also be used as an attribute, see [Terminology](https://docs.rs/clap/latest/clap/_derive/index.html#terminology) for syntax." so it can be used in the derive API but I think I'll go with the two `Parsers`

**Review by dpc — APPROVED** [review]

**Comment by dpc** [comment]
@NicolaLS I think it's better to land already, and please follow up with the command line parsing improvements in another PR.
